SCAutolib/models/file.py

```diff
@@ -0,0 +1,104 @@
+"""
+from configparser import ConfigParser
+from pathlib import Path
+from typing import Union
+
+from SCAutolib import logger
+
+
+class File:
+    """
+    This class defines an interface for generic operations on config files
+
+    create: create content of config file based on template file
+    set:    modify content of config files, add keys or sections if necessary
+    save:   save config file
+    clean:  remove config file
+    """
+    _conf_file = None
+    _template = None
+    _default_parser = None
+
+    def __init__(self, filepath: Union[str, Path], template: str = None):
+        """
+        Init of File
+
+        :param filepath: Path of config file
+        :type filepath: str of pathlib.Path
+        :param template: Path of template file
+        :type template: str
+        """
+        self._conf_file = Path(filepath)
+        if template is not None:
+            self._template = Path(template)
+
+    def create(self):
+        """
+        Populate internal parser object with content based on template.
+        """
+        if self._conf_file.exists():
+            logger.warning(f"Create error: {self._conf_file} already exists.")
+            raise FileExistsError(f'{self._conf_file} already exists')
+        else:
+            self._default_parser = ConfigParser()
+            self._default_parser.optionxform = str
+            with self._template.open() as t:
+                self._default_parser.read_file(t)
+
+    def set(self, key: str, value: Union[int, str, bool], section: str = None):
+        """
+        Modify value in config file.
+
+        :param key: value for this key will be updated
+        :type key: str
+        :param value: new value to be stored in [section][key] of config file
+        :type value: int or str or bool
+        :param section: section of config file that will be modified
+        :type section: str
+        """
+        if self._default_parser is None:
+            self._default_parser = ConfigParser()
+            self._default_parser.optionxform = str
+            with self._conf_file.open() as config:
+                self._default_parser.read_file(config)
+        if not self._default_parser.has_section(section):
+            logger.warning(f"Section {section} not present.")
+            logger.info(f"Adding section {section}.")
+            self._default_parser.add_section(section)
+
+        previous = self._default_parser.get(section, key, fallback="Not set")
+
+        self._default_parser.set(section, key, value)
+        logger.info(f"Value is changed.")
+        logger.debug(f"Old value in section [{section}] {key}={previous}")
+        logger.debug(f"New value in section [{section}] {key}={value}")
+
+    def save(self):
+        """
+        Save content of config file stored in parser object to config file.
+        """
+        with self._conf_file.open("w") as config:
+            self._default_parser.write(config)
```

Comment:
Does that mean if we do only set() the content won't be written to the file? If so it is probably a good idea to move this code to set()

Reply:
Modifying content and writing it to a file was separated intentionally. Reasons are following:
1) unit tests. If unit tests write files (system config files) they may harm system and it complicates unit testing.
2) usability - this way, user can create a content and save it to separate (temporary) file if needed or to do multiple changes before writing file. 

Note, that this approach was implemented through whole File class and child classes and that restarting services, for example sssd is also separated from creation and modification of files and current plan is that controller would take care of it.

Reply:
1. Most of the class testing can be done on non system config files(things like appending, removing, with section, without section, etc) and it would be better to test those things in unittests. Modifying system config files can be done only in our CI for example, and excluded by default if somone runs them locally with  something like @unittest.skipUnless(condition, reason)
2. I think this is a bit over complicating things, I feel like after an object is created, set() should just simply modify the file, but we can discuss it further for pros/cons.

Reply:
@inikolcev  Let me comment on this `save` method. It is a method of the model. The user wouldn't communicate directly with the model. This communication should be done via the controller. So, the user wouldn't need to call `save` method after calling set to, he would call just `controller.set()` on any other method name. And this method would modify the content and then save the file.

This behavior is defined by the pattern we are using. Models should provide atomic functionality, which is just changing some value, just save, and etc.